smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/PaginationGenerator.java

```diff
@@ -0,0 +1,199 @@
+            writer.write("client: $L | $L;", nonModularServiceName, service.getName());
+        });
+    }
+
+    private void writePager() {
+        String serviceTypeName = serviceSymbol.getName();
+        String inputTypeName = inputSymbol.getName();
+        String outputTypeName = outputSymbol.getName();
+        String inputTokenName = paginatedInfo.getInputTokenMember().getMemberName();
+        String outputTokenName = paginatedInfo.getOutputTokenMember().getMemberName();
+
+        writer.openBlock(
+                "export async function* $LPaginate(config: $L, input: $L, ...additionalArguments: any): Paginator<$L>{",
+                "}",  methodName, paginationType, inputTypeName, outputTypeName, () -> {
+            writer.write("let token: string | undefined = config.startingToken || '';");
+
+            writer.write("let hasNext = true;");
+            writer.write("let page:$L;", outputTypeName);
+            writer.openBlock("while (hasNext) {", "}", () -> {
+                writer.write("input[\"$L\"] = token;", inputTokenName);
+                if (paginatedInfo.getPageSizeMember().isPresent()) {
+                    String pageSize = paginatedInfo.getPageSizeMember().get().getMemberName();
+                    writer.write("input[\"$L\"] = config.pageSize;", pageSize);
+                }
+
+                writer.openBlock("if (config.client instanceof $L) {", "}", nonModularServiceName, () -> {
+                    writer.write("page = await makePagedRequest(config.client, input, ...additionalArguments);");
+                });
+                writer.openBlock("else if (config.client instanceof $L) {", "}", serviceTypeName, () -> {
+                    writer.write(" page = await makePagedClientRequest(config.client, input, ...additionalArguments);");
+                });
+                writer.openBlock("else {", "}", () -> {
+                    writer.write(" throw new Error(\"Invalid client, expected $L | $L\");",
+                            nonModularServiceName, serviceTypeName);
+                });
+
+                writer.write("yield page;");
+                if (outputTokenName.contains(".")) {
+                    // Smithy allows one level indexing (ex. 'bucket.outputToken').
+                    String[] outputIndex = outputTokenName.split("\\.");
+                    writer.write("token = page[\"$L\"][\"$L\"];", outputIndex[0], outputIndex[1]);
+                } else {
+                    writer.write("token = page[\"$L\"];", outputTokenName);
+                }
+
+                writer.write("hasNext = !!(token);");
+            });
+
+            writer.write("// @ts-ignore");
+            writer.write("return undefined;");
+        });
+    }
+
+
+    /**
+     * Paginated command that calls Command({...}) under the hood. This is meant for server side environments and
+     * exposes the entire service.
+     */
+    private void writeRequest() {
```

Comment:
Maybe these should be `writeMethodRequest` and `writeCommandRequest`? This clarifies what the code itself is doing and not who the expected consumer is.